### PR TITLE
Added preprocessor and Changed Apache to GPL

### PR DESCRIPTION
**Pull request by firefish111**
# Done
@DynamicSquid ***done***

### CODE REVIEW

File: src/file_extraction.h [hunk 1]
```diff
@@ -37,6 +38,23 @@ void ExtractFile(const char* fileName)
...
+        std::stringstream tik;
+        tik << "./dusk_pkgs/" << tokens[a].value.cstr();
+        codeFile = Read(tik.str().c_str());
+        while (getline(codeFile, fileLine))
+        {
+            night::array<Token> temp = Lexer(fileLine.c_str());
+            for (int a = 0; a < temp.length(); ++a)
+                tokens.push_back(temp[a]);
+        }
+        }
```
Comment (DynamicSquid) [inline src/file_extraction.h:56]: what is this indentation

Comment (firefish111) [reply src/file_extraction.h:56]: @DynamicSquid welp, looks like vs code is absolute s*** then

**Comment by DynamicSquid** [comment]
k I'll definitely take a look at it. might create a new release for that even. but I'll put it on hold for now since I really need to get v3.1 out there

File: src/file_extraction.h (hunk 1 above)
Comment (DynamicSquid) [reply src/file_extraction.h]: wdym its just copy and paste?

**Comment by DynamicSquid** [comment]
oof, your tests failed. lemme take a look

File: src/file_extraction.h (hunk 1 above)
Comment (firefish111) [reply src/file_extraction.h]: No, I have VIM keymaps. `<<` and `>>` are not working as expected on VSCode

**Comment by DynamicSquid** [comment]
could you merge the latest commit with your fork?

**Comment by firefish111** [comment]
@DynamicSquid What? I can't merge, only you can.
~~(we can make a night-dev organisation so all the contributors are together as one)~~

File: src/file_extraction.h (hunk 1 above)
Comment (DynamicSquid) [reply src/file_extraction.h]: oh okay. could you fix it though?

**Comment by DynamicSquid** [comment]
no your fork merge with this. you have to pick up the latest commits on the .yml file

**Comment by firefish111** [comment]
Will do

**Comment by firefish111** [comment]
@DynamicSquid Done

**Comment by DynamicSquid** [comment]
Cool!

**Comment by firefish111** [comment]
Alrighty! We can merge now!
~~Also would you like me to create that night-dev organisation~~

**Comment by DynamicSquid** [comment]
no and no. I want this to be part of a releases, but I don't know which yet. maybe 3.1.1 or 3.0.2

**Comment by firefish111** [comment]
@DynamicSquid Weell maybe 3.3 since it only really works in conjunction with dusk

**Comment by DynamicSquid** [comment]
really? why is that?

**Comment by DynamicSquid** [comment]
if you make a separate PR with the license, i'd be happy to merge that right now

**Comment by firefish111** [comment]
@DynamicSquid look at the code
image

**Comment by DynamicSquid** [comment]
ah

**Comment by DynamicSquid** [comment]
okay 3.3.then

**Comment by firefish111** [comment]
When is the estimated release data of 3.3?

**Comment by DynamicSquid** [comment]
August 31st 2020

**Comment by firefish111** [comment]
That was four days ago

**Comment by DynamicSquid** [comment]
Yup

**Comment by firefish111** [comment]
So it was mweant to be released then?!?! Well what is the now estimated release, Halloween?

**Comment by DynamicSquid** [comment]
Well really, the major thing holding me back is v3.1 Once that's done, then every else should be smooth sailing so probs like 1/2 weeks after I release 3.1 then 3.3 will come out. So the big question is when do I get v3.1 done? But no, it's not gonna be on Halloween lol. I was thinking maybe sometime in November

**Comment by DynamicSquid** [comment]
jk lol . either tmr or the 11th

**Comment by DynamicSquid** [comment]
ight imma take a break now

**Comment by firefish111** [comment]
Bye!

**Comment by DynamicSquid** [comment]
update: array return values work now (there might be 100000000000000000000000000000000000000000000 hidden bugs but idk at least it works)

now I have to do array params

**Comment by firefish111** [comment]
@DynamicSquid Good!
###### ~~The world would be a much better place if people didn't ping each other late at night, but I guess that makes me a hypocrite~~

**Comment by firefish111** [comment]
@DynamicSquid Great news! Added the `extern` command!

**Comment by DynamicSquid** [comment]
@firefish111 Quick question, when you import a package using the package manager, will the file structure look like this?

```
src
dusk_pkgs
```

Or this?

```
src
  dusk_pkgs
```

**Comment by firefish111** [comment]
If the source files are in src, then the latter. @DynamicSquid

**Comment by DynamicSquid** [comment]
so the second one?

**Comment by firefish111** [comment]
@DynamicSquid ~~do you not know what latter means~~ yes

**Comment by DynamicSquid** [comment]
wait but should we have it outside the `src` file?

**Comment by firefish111** [comment]
@DynamicSquid Well if we did that repl.it's infrastructure will collapse
~~We are imitating nodejs remember~~

**Comment by firefish111** [comment]
@DynamicSquid I'm sorry, but added `use` and `include`. `import` sounds weird to me; hope you understand.
if you want to use import you have to change it in lexer.h and file-extraction.h

**Comment by CodeLongAndProsper90** [comment]
I like the syntax `please add {{module}}, ok?;`

**Comment by firefish111** [comment]
@CodeLongAndProsper90 I hope you are joking, if you aren't, then go make rudeLang

**Comment by CodeLongAndProsper90** [comment]
> rudeLang

What? That was a polite way of asking the compiler.

**Comment by DynamicSquid** [comment]
@firefish111, @CodeLongAndProsper90 be nice to your ide/compiler

https://img.devrant.com/devrant/rant/r_2075836_ZBpcZ.jpg

**Comment by firefish111** [comment]
@DynamicSquid lol xd
~~also lol I have both on my windows dualboot~~
image

**Comment by firefish111** [comment]
@DynamicSquid Also the 
image
It was the VM for Night and Slice. I write a compiler for both, that compiles to the same virtual machine, that runs the code. You can keep the interpreter code, so we get like python, compile a script. Also during v4, shall we uh, um, add modules, so we can sort code? so like
```
mod lala
{
  null squid()
  {
     print("squid cool");
  }
}
lala:squid();
// should output squid cool
```

**Comment by CodeLongAndProsper90** [comment]
Another thing, though I'm not the BDFL here, perhaps `none` instead of null? Because a null squid does not exist.

**Comment by firefish111** [comment]
@CodeLongAndProsper90 Personally I prefer `nil` but, y'know I didn't make this...
~~also you are becoming the BDFL, so be careful~~

**Comment by DynamicSquid** [comment]
@firefish111 I think I might add something like C structs in `v4`... I don't know yet. I just want a lightweight encapsulation method.  I don't know about modules yet, but that would be cool

**Comment by firefish111** [comment]
@DynamicSquid VS, by miles, Eclipse is godawful
Also, structs and unions would be a great idea. (in case you don't know what a union is, basically a struct but only one member can have a value at any given time, setting another member erases the value of the first.)

**Comment by DynamicSquid** [comment]
@firefish111 yeah ik what a union is I'm not that dumb lol. also yeah, unions would be a good idea

**Comment by firefish111** [comment]
@DynamicSquid So when is v3.2 gonna be released, and this pr merged?

**Comment by DynamicSquid** [comment]
Have you tested this pr? Does it and dusk work?

**Comment by firefish111** [comment]
@DynamicSquid Well dusk works, that's for certain. Not sure about this pr though, because this is an old version that still throws errors, but they will be merged away.

**Comment by DynamicSquid** [comment]
it works it works it works yeys sys ye si fdg it do goh oI fix bgsf ookay this is going good

**Pull request closed without merging**